python/ray/dataframe/dataframe.py

```diff
@@ -555,16 +621,188 @@ def expanding(self, min_periods=1, freq=None, center=False, axis=0):
+                    last_row_df = pd.DataFrame(
+                        [df.iloc[0, :] for df in ray.get(new_df._df[1:])]
+                    )
+                last_row_df.fillna(value=value, method=method, axis=axis,
+                                   inplace=True, limit=limit,
+                                   downcast=downcast, **kwargs)
+                if is_ffill:
+                    new_df._df[1:] = [
+                        _deploy_func.remote(fill_in_part, new_df._df[i + 1],
+                                            last_row_df.iloc[i, :])
+                        for i in range(len(self._df) - 1)
+                    ]
+                else:
+                    new_df._df[:-1] = [
+                        _deploy_func.remote(fill_in_part, new_df._df[i],
+                                            last_row_df.iloc[i])
+                        for i in range(len(self._df) - 1)
+                    ]
+
+            if limit is not None:
+                # need to do a seperate fillna for each column seperatley
+                # count how many N/As there are in each partition
+                nan_counts = ray.get(self._map_partitions(
+                    lambda df: df.isnull().sum()
+                )._df)
+                identity = self._map_partitions(
+                    lambda df: df.copy()
+                )
+                col_len = len(self.columns)
+                stopping_points = [-1] * col_len
+                left_over = [limit] * col_len
+                current_count = pd.Series([0] * col_len)
+
+                count_idx_gen = range(len(nan_counts))
+
+                if is_bfill:
+                    count_idx_gen = range(-1, -1 * (len(nan_counts) + 1), -1)
+
+                for i in count_idx_gen:
+                    current_count += nan_counts[i]
+                    for j in range(col_len):
+                        if stopping_points[j] == -1\
+                           and current_count[j] >= limit:
+                            stopping_points[j] = i
+                            left_over[j] = limit - (
+                                current_count - nan_counts[i]
+                            )[j]
+
+                @ray.remote
+                def copy_part_col(i, j, new_df, identity):
+                    """
+                    Args:
+                        i: The column index to copy
+                        j: The df partition to work on
+                        new_df: The ray df w/ non-limited fills
+                        identity: The ray df w/ no fills
+
+                    Returns:
+                        A pandas df with it's ith column replaced with
+                        the original column
+                    """
+                    n_df = ray.get(new_df._df[j]).copy()
+                    i_df = ray.get(identity._df[j]).copy()
+                    n_df[[n_df.columns[i]]] = i_df[[i_df.columns[i]]]
+                    return n_df
+
+                @ray.remote
```

Comment:
@robertnishihara Should we allow inner functions to be defined as remote?

Reply:
Hmm... it should work.. but there is overhead every time a remote function is defined. So every time the `fillna` method is called, this will store the remote function definition in Redis and broadcast the definition to all of the workers. Whereas if it defined outside then that will only happen once.

How hard would it be to define this as a helper function somewhere outside of the method invocation?

Reply:
It should not be too hard to do that. I define it outside.